make rpc multiprocess test to use spawn instead of fork

zhaojuanmao opened the pull request:
Summary:
1. current fork unit tests did not catch rpc exit crash issue, but spawn caught it. most of use cases also uses spawn,
so changing multiprocessing start method from fork to spawn

2. it did not change fork to spawn for c10d tests for now -- if we need to make changes for c10d, it may need another PR as c10d has a lot of unit tests, may need some more changes to make spawn work for test_c10d

3. fix the real segment exit issue caused by python rpc handler temporarily by removing module_ global variable, separate diff D17097999 is to change python rpc handler to be singleton class for deterministic destruction order of variables

Test Plan:
1. use fork without this diff, rpc tests passed even there is process exit crash issue
2. use spawn in this diff, rpc tests caught crash and failed;
3. use spawn with fixing python rpc handler, rpc tests passed;
4. c10d tests passed on CPU and GPU hosts

Differential Revision: D17086007

satgera commented on test/common_distributed.py:
```diff
@@ -129,26 +129,39 @@ def setUp(self):
-        getattr(self, self.id().split(".")[2])()
+        # We're retrieving a corresponding test and executing it.
+        getattr(self, test_name)()
         sys.exit(0)
```
Why exit(zero) explicitly shouldn't this return what is returned from the test ?

zhaojuanmao replied:
@satgera I do not know the historic reason, but I guess put it here explicitly is no harm? if sys exit with other codes earlier, sys.exit(0) will not be executed; if sys.exit(0) is executed here, then it is fine as well

xush6528 replied:
@satgera 

I guess it's because the sub-process was forked, so they would run `self.tearDown()`, because the TestCase object was hold in a "TestRunner" object.

and sys.exit() was for skipping `self.tearDown()`.

However, with this PR, the TestCase object is not loaded in a test runner, `self.tearDown()` will not be run.

So I think it's safe to remove `sys.exit(0)` now.

xush6528 commented on test/test_rpc.py:
```diff
@@ -90,7 +92,16 @@ def wrapper(self):
+    @classmethod
+    def setUpClass(cls):
+        multiprocessing.set_start_method("spawn")
+        for attr in dir(cls):
```
Can you call `super(RpcTest, self). setUpClass()` instead, to avoid duplicating code?

zhaojuanmao replied:
will remove sys.exit(0) after c10d use spawn, right now c10d is still using fork

zhaojuanmao committed:
make rpc multiprocess test to use spawn instead of fork (#25558)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/25558

1. current fork unit tests did not catch rpc exit crash issue, but spawn caught it. Pyper also uses spawn,
so changing multiprocessing start method from fork to spawn

2. it did not change fork to spawn for c10d tests for now -- if we need to make changes for c10d, it may need another PR as c10d has a lot of unit tests, may need some more changes to make spawn work for test_c10d

3. fix the real segment exit issue caused by python rpc handler temporarily by removing module_ global variable, separate diff D17097999 is to change python rpc handler to be singletion class for deterministic destruction order of variables

Test Plan:
1. use fork without this diff, rpc tests passed even there is process exit crash issue
2. use spawn in this diff, rpc tests caught crash and failed;
3. use spawn with fixing python rpc handler, rpc tests passed;
4. c10d tests passed on CPU and GPU hosts

Reviewed By: xush6528

Differential Revision: D17086007

fbshipit-source-id: b1fc031bf226d6f2e51106455b9ef86be95177f3

pietern commented on test/common_distributed.py:
```diff
@@ -129,26 +129,39 @@ def setUp(self):
         self.rank = self.MAIN_PROCESS_RANK
-        self.file = tempfile.NamedTemporaryFile(delete=False)
+        # spawn can not pass file handle, so pass file name here
+        self.file = tempfile.NamedTemporaryFile(delete=False).name
```
This is no longer a file, but a file name. Can you rename the var to reflect that?

pietern commented on test/test_rpc.py:
```diff
@@ -90,7 +92,13 @@ def wrapper(self):
+    @classmethod
+    def setUpClass(cls):
+        multiprocessing.set_start_method("spawn")
+        super(RpcTest, cls). setUpClass()
```
Initialization of the super class is usually done first. Otherwise it could override whatever you're executing here.

Is there a reason for the reversed order here?

pietern reviewed (commented):
@zhaojuanmao A few more comments.

Did you check the impact of using spawn on the test runtime?

xush6528 reviewed (changes requested):
For Pieter's comments

zhaojuanmao commented:
closing this one, #25656 is new one using ghexport
